nat-lab/tests/test_direct_connection.py

```diff
@@ -76,90 +92,223 @@
+    ),
 ]
 
 
-@pytest.mark.asyncio
-@pytest.mark.timeout(150)
-@pytest.mark.parametrize(
-    "endpoint_providers, client1_type, client2_type, reflexive_ip",
-    UHP_conn_client_types,
-)
-async def test_direct_working_paths(
-    endpoint_providers, client1_type, client2_type, reflexive_ip
-) -> None:
-    async with AsyncExitStack() as exit_stack:
-        api = API()
-        (alpha, beta) = api.default_config_two_nodes()
+@dataclass
+class NodeWithMeshConnection:
+    node: Node
+    client: telio.Client
+    conn: Connection
+    conn_track: ConnectionTracker
+    conn_gw: Optional[Connection]
+
+    def _init_(self, node, client, conn, conn_track, conn_gw=None):
+        self.node = node
+        self.client = client
+        self.conn = conn
+        self.conn_track = conn_track
+        self.conn_gw = conn_gw
+
+
+@asynccontextmanager
+async def new_connections_with_mesh_clients(
+    exit_stack: AsyncExitStack,
+    client1_type: ConnectionTag,
+    endpoint_providers_1: List[str],
+    client2_type: ConnectionTag,
+    endpoint_providers_2: List[str],
+    client3_type: Optional[ConnectionTag] = None,
+    endpoint_providers_3: Optional[List[str]] = None,
+) -> AsyncIterator[
+    Tuple[
+        NodeWithMeshConnection, NodeWithMeshConnection, Optional[NodeWithMeshConnection]
+    ]
+]:
+    api = API()
+
+    (alpha, beta, gamma) = api.default_config_three_nodes()
 
-        alpha_connection = await exit_stack.enter_async_context(
-            new_connection_by_tag(client1_type)
+    (
+        alpha_conn,
+        alpha_conn_gw,
+        alpha_conn_tracker,
+    ) = await exit_stack.enter_async_context(
+        new_connection_with_tracker_and_gw(
+            client1_type,
+            generate_connection_tracker_config(
+                client1_type,
+                derp_0_limits=ConnectionLimits(0, 0),
+                derp_1_limits=ConnectionLimits(1, 3),
+                derp_2_limits=ConnectionLimits(0, 3),
+                derp_3_limits=ConnectionLimits(0, 3),
+                ping_limits=ConnectionLimits(0, 5),
+            ),
         )
+    )
 
-        beta_connection = await exit_stack.enter_async_context(
-            new_connection_by_tag(client2_type)
+    (
+        beta_conn,
+        beta_conn_gw,
+        beta_conn_tracker,
+    ) = await exit_stack.enter_async_context(
+        new_connection_with_tracker_and_gw(
+            client2_type,
+            generate_connection_tracker_config(
+                client2_type,
+                derp_0_limits=ConnectionLimits(0, 0),
+                derp_1_limits=ConnectionLimits(1, 3),
+                derp_2_limits=ConnectionLimits(0, 3),
+                derp_3_limits=ConnectionLimits(0, 3),
+                ping_limits=ConnectionLimits(0, 5),
+            ),
+        )
+    )
+
+    alpha_client = await exit_stack.enter_async_context(
+        telio.Client(
+            alpha_conn,
+            alpha,
+            telio.AdapterType.BoringTun,
```

Comment:
Is it required to use BoringTun? On another tests the adapter is ommited to telio.Client constructor, leaving it as AdapterType.Default